docs/spec/interfaces.md

```diff
@@ -0,0 +1,93 @@
+# Interface Contracts
+
+<!-- This image is sourced from https://drive.google.com/open?id=1Mgn0Oz85M9W1Pwfnw3u0OhZmIIxSxcM3lZUEZyPCoB0 -->
+![Interface Contracts Overview](images/interface-contracts-overview.svg)
+
+---
+
+## Subscribable
+
+A **Subscribable** resource will emit events that a _Subscription_ can direct
+to a _Targetable_ or _Sinkable_ resource.
+
+### Control Plane
+
+A **Subscribable** resource may be referenced in the _from_ section of a
+_Subscription_. The **Subscribable** resource MUST expose a
+_subscribable.channelable_ field (an ObjectReference) in its _status_ section.
+This _channelable_ field MUST be _Channelable_, and _channel_ MAY refer back to
+the **Subscribable** resource.
+
+### Data Plane
+
+A **Subscribable** resource is an event producer or forwarder, events that it
+produces or forwards are delivered via its _status.channel_ resource.
```

Comment:
I just realized that if we have multiple _Subscribable_ resources which point at the same _Channelable_, the Subscription controller will need some sort of global knowledge to build the list of `spec.channelable` resolved endpoints. (I.e. if I discover that a subscription has been deleted, I can't necessarily remove it from the Channel if there was a duplicate subscription elsewhere. Maybe we already have that solved somehow?)

@vaikas-google @grantr @n3wscott who have been looking at the Subscribable interface.

Reply:
I'm not sure if we're talking about a same thing, but I had opened an issue about a week ago on this here:
https://github.com/knative/pkg/issues/96

with a possible solution.

Reply:
I think Channels should be the only subscribable resources. N:1 relationships can exist between Sources and Channels, and that creates issues beyond the cascading deletion. No single Source is necessarily the owner of a Channel. So, what does it mean to subscribe to a Source whose target Channel is also a target of other Sources? The subscriber would be receiving events from other Sources in addition to the one to which it intends to subscribe. More generally, the main role of a "channel" (or topic, etc) in a messaging system is to decouple publishers from subscribers. We lose that benefit if we bypass the channel and subscribe directly to the Source.

Also it was my understanding that only Channel (and a future ClusterChannel) would be subscribable based on this thread: 
https://github.com/knative/eventing/pull/421#issuecomment-421092562

Reply:
> We lose that benefit if we bypass the channel and subscribe directly to the Source.

I think there might be some confusion about what's happened in the repo in the last week or two. That's understandable given that you've been busy elsewhere :) Here's a summary:

- We've pushed the authoritative list of subscribers into the Channel spec as a `subscribers` array (specified by [the `Channelable` duck type](https://github.com/knative/pkg/blob/49f84a651e102bb36fba49e9fde035c71e51e3dd/apis/duck/v1alpha1/channelable_types.go#L27-L33), naming suggestions welcome). This is equivalent to the previous Subscription object.
-  **Sources do not have a `subscribers` array.** Sources contain a reference to a Channel in their status via the [the `Subscribable` duck type](https://github.com/knative/pkg/blob/49f84a651e102bb36fba49e9fde035c71e51e3dd/apis/duck/v1alpha1/subscribable_types.go#L28-L37).
- We've pushed the resolution of `Subscribable` resources, previously the responsibility of Flow, into Subscriptions. The new Subscription object takes the place of Flow and Flow goes away.

Does that make sense? I'll avoid speaking for anyone else, but the scenarios I've discussed and envisioned for subscribing to a Source do not bypass channels at all. **All Sources and other Subscribables must reference a Channel** that implements the decoupling behavior you describe. The subscription controller is responsible for resolving the Subscription's reference to the Source into a reference to its Channel and updating the Channel `subscribers` array to include the Subscription's `call` and `result`.

It's possible for a Source to not be Subscribable: this would be a Source without a Channel reference in its status. That Source would not be usable in a Subscription's `from` field, or more accurately, the Subscription's `Ready` condition would be `false` until the Source referenced a Channel in its status.

Stepping back from implementation and into user experience for a moment, my vision for the steps the user takes to create a Source and subscribe to it is this:
1. User creates a Source.
2. User creates a Subscription with the Source as the `from` reference.

Unbeknownst to the user, the following process occurs:
1. The Source controller creates a Channel.
2. The Source controller adds a reference to that Channel in the Source's status.
3. The Subscription controller resolves the Channel in the Source's status and updates its `subscribers` array.

IMO, this is the reason for the Subscribable duck type. It allows the user to describe the flow between two logical resources without needing to know that there is in fact a Channel object handling transport between them.

@markfisher Can you describe the user experience you imagine for a user attempting to create a source and subscribe to its output? What steps do they take?

> what does it mean to subscribe to a Source whose target Channel is also a target of other Sources?

I consider this an edge case, since most of the scenarios we've envisioned have a Source creating its own Channel as part of the provisioning process, but I'd love to hear what scenarios you're thinking of. I imagine that if a user did this, it would be intentional and they would expect to get exactly the behavior that you describe.

Reply:
Thanks for the explanation @grantr! and sorry for losing track of things last week. Are there any docs, Github issues, etc that provide more of the context for these decisions, or is it just in PRs?

I view the Source as a "producer", and in most messaging systems a Consumer doesn't subscribe directly to a Producer. Instead a Producer publishes events to a Channel, and the Consumer subscribes to the Channel. I consider "decoupling" to include the intentional prevention of that direct awareness. It should be possible to swap/add/remove Producers and Consumers without them impacting each other.

While I would agree that Flow could've used some changes, I much prefer a model where CRDs are layered such that each Controller handles the Resource type its reconciling in a well-defined way without conditional logic determining at which level it's operating. In fact, that was one area where I thought there was unanimous agreement not long ago when Flow was first introduced - because at that level, users didn't want to have to think about the Channel, but we still agreed that the Channel was providing value.

The lower layer building blocks would be Source/Channel/Subscription, and the higher level would be something (was Flow) that generates and/or connects them. IMO it complicates things to have this dual role of Subscription - sometimes being at Channel level and sometimes playing a higher level role of finding the Channel indirectly via a Source. We (riff) will definitely be building "chains" of Functions, and we were hoping that Flow - or something that would replace it but still be at a higher level - would evolve to support multiple "callables" (to support: `source | fn1 | fn2 | sink`), and we'll also want to support nesting of these higher level chains to create graph "flows" as well. So a higher level CRD would be needed for those types of patterns whether that's defined in Knative or riff.

I also think CLIs and other UIs would eventually be providing the actual user experience for any platform/product built on top of Knative, and for that reason I think it's a mistake to consider the YAML itself as the user experience. These convenient shortcuts lead to a confusing architecture in my opinion.

Finally, I still don't understand how this avoids the problem of multiple Sources targeting the same Channel. It seems counter-intuitive that a user intending to subscribe to only one of those Sources would inadvertently receive events from other Sources as well. It's much clearer to have Channels as the only subscribable resource, perhaps with a configuration option to make a Channel "exclusive" from the Source perspective.

Reply:
> sorry for losing track of things last week. Are there any docs, Github issues, etc that provide more of the context for these decisions, or is it just in PRs?

No need to apologize! The context has been in PRs, with some of it reaching the specification doc. This PR is attempting to move that context to a more consumable form.

> in most messaging systems a Consumer doesn't subscribe directly to a Producer.

> at that [Flow] level, users didn't want to have to think about the Channel

I'm a little confused about what exactly you're arguing in favor of, but I want to understand. 😸   You don't want Consumers subscribing to Producers, but you support Flow allowing (what seems like) exactly that.

I think we want an API that matches the concepts users have in their heads, and that may not be what messaging systems implement. Based on your statements above I think we agree that users will not naturally understand that "a Source is a Producer and Consumers don't subscribe directly to Producers." Instead users have goals like "events come from A and I want them to go to B." I think you are arguing that this goal _should be expressible_, but that it should not be implemented at the lowest layer. Is that accurate?

> The lower layer building blocks would be Source/Channel/Subscription, and the higher level would be something (was Flow) that generates and/or connects them.

As I said above, these lower layer concepts are still in the spec, except that the previous concept of Subscription has been moved into Channel as a subscriber array. We then folded some of the features formerly implemented by Flow into Subscription, which may have been a source of confusion.

It sounds like you want Flow (or the capabilities of Flow) to be the preferred user interface, with the lower layer an internal detail. Would it help your understanding if the proposed Subscription kind was renamed to Flow? Because it's basically Flow. :grin:

Here's the approximate thought process that led to the current Subscription proposal:

1. Subscriptions will be easier for Channel implementers to implement if they are aggregated into a resolved subscriber list inside the Channel spec.
1. If we have this list, then Subscription as the authoritative record of the resolved links between Channels is redundant and no longer needed.
1. Instead of removing Subscription entirely, it can be a common implementation of Channel resolution.
1. Now that Subscription handles Channel resolution, it's very similar to Flow. Is it worth keeping Flow as a separate object?

I think this change may have been easier to process if we had just removed Subscription and let Flow keep its resolution capabilities. (cc @vaikas-google)

> I think it's a mistake to consider the YAML itself as the user experience.

I tend to disagree with this. The YAML is _one of_ the user experiences, and it's an important one because it's what users see when everything blows up and they need to debug.

Reply:
> I'm a little confused about what exactly you're arguing in favor of, but I want to understand. 😸 You don't want Consumers subscribing to Producers, but you support Flow allowing (what seems like) exactly that.

I would prefer a model that does not support the direct Source->Function at the lowest level of CRD (should always be Source->Channel->Function), but I would be okay with a higher level CRD that creates Channels implicitly if they don't yet exist, perhaps even Channels that are "anonymous" and/or "exclusive". For example, with riff we are considering supporting the DSL that we've used for Spring Cloud Data Flow which allows syntax like: `source | function1 | function2 | sink`. Even when using that DSL, the consuming Functions and Apps are always subscribing to Channels, but those Channels could be created by the Controller that reconciles the higher level CRD. The pipe symbol is succinct, but it still serves as a reminder that there is always a Channel.

> I think we want an API that matches the concepts users have in their heads, and that may not be what messaging systems implement. Based on your statements above I think we agree that users will not naturally understand that "a Source is a Producer and Consumers don't subscribe directly to Producers." Instead users have goals like "events come from A and I want them to go to B." I think you are arguing that this goal should be expressible, but that it should not be implemented at the lowest layer. Is that accurate?

I think this is describing the exact reason we need 2 distinct layers with a well-defined boundary. Some users (and especially implementors building on top of Knative, such as riff) *do* have messaging system concepts in their heads. I have supported many users and customers over the years who absolutely think in terms of messaging system concepts and enterprise integration patterns in the solutions they are building. So the bottom layer should match those concepts in my opinion. A layer above that can provide the "events come from A and I want them to go to B" case, similar to the DSL I mentioned above or some alternative representation (including visual UIs), but that layer itself should translate into creation/wiring of the lower level resources.

> It sounds like you want Flow (or the capabilities of Flow) to be the preferred user interface, with the lower layer an internal detail.

I actually view only the lower layer as essential for Knative itself. It could be left to the platforms that build on top of Knative to provide the higher layer experiences. That said, if the wider community can agree on the higher level model, that's even better.

> Would it help your understanding if the proposed Subscription kind was renamed to Flow? Because it's basically Flow.

I was not a huge fan of the term "Flow" for the role it played before. However, I think those thinking at the level of messaging system concepts will associate Subscription directly with the Channel. This is obviously the point I'm hung up on myself. If I had to choose between the 2, I would go with "Flow", but I don't think simply renaming is sufficient to alleviate my concerns. Hopefully I'm making it clear why I find value in maintaining both levels.

Frankly, I'm a little lost in the 4 points of the thought process. I do understand the goal of "one stop shopping" for Channel implementors, but that feels like an unworthy tradeoff for the complexity it adds when actually registering a subscribing Function to a Channel (but maybe I'm missing something about how that works in this model).

> The YAML is one of the user experiences, and it's an important one because it's what users see when everything blows up and they need to debug.

I agree 100%, and I should have been much more clear in my comment. I very much view YAML wrangling as *one* of the user experiences, and perhaps even the most important one. My concern is that it feels like the YAML is being considered as the *exclusive* user experience, and that a clever shortcut is being considered as a way to make a relatively flat model do things that should be spread across layers. I think a well-defined layered architecture of CRDs where verbosity may increase as you drop down each layer is even more beneficial for debugging. The same applies to K8s as a whole. For example, I rarely apply YAML at the Pod level, but I often read configuration and status at the Pod level.

Hopefully I've clarified my concerns a bit here. I'd like to dive into more of the detail behind those 4 points of the thought process so I understand better. Slack might be a better venue as you pointed out there earlier today. Thanks @grantr !

Reply:
As someone just wading into this discussion now, and at the risk of oversimplifying, I'll describe how I see the details of this specific discussion.

There are two use-cases described here:

1. In this design a user can explicitly create a Channel. She can then create a Source that sends events to that Channel and a Subscription that links an endpoint with events from that Channel.

1. Alternatively, she can create a Source without specifying a Channel and a Subscription that links an endpoint with events from that Source. A Channel is implicitly created and used in this scenario. A Source is Subscribable (and will implicitly create a Channel) to support this usage.

It looks like everyone agrees that use-case 1 is desirable. The discussion is centered around whether use-case 2 should be achievable using the same abstraction, or whether a new Object (Flow?) should be introduced to represent connecting a Source to an endpoint. And that decision has implications for other parts of the design.

To prevent confusing things from happening, a Source may need to "remember" if it was created with an explicit or implicit (and unique) Channel. And, when linking a Source to an endpoint via a Subscription, the Subscription may need to ensure the Source has an implicit Channel. Otherwise you can end up in scenarios described above where the user creates a Subscription to a Source but ends up getting events from other Sources.

Reply:
I'm a bit confused that the Source controller is supposed to create a Channel, in case the Subscription is directly referring to a Source. Does the Source controller really have the information about the best way to bring an event to a Subscriber? For the Flow I had in mind that there might be additional attributes to describe the quality of the event flow, e.g. at-least-once semantics. Looking at Channel creation just from the source perspective appears to me a bit odd. Even if not the Source controller is creating the Channel, deciding about the Channel when the Source is created and not when the Subscription happens, seems strange.

Perhaps I am missing something. Do you have an example for which a Source could make a good choice for the default Channel?